### ISSUE TITLE
Support Adaptive Max Gradient Norm / Clipping

### ISSUE THREAD

**Issue by PetrochukM** [pytorch/pytorch#12189]
## 🚀 Feature
This feature is to use a moving windowed median or exponential moving average for gradient clipping and normalization. 

Additionally, it'd be nice to support batch skipping if the gradient norm is above a threshold which is usually an indication of an imminent gradient explosion.

## Motivation

The threshold for gradient clipping is a hyperparameter that needs to be set uniquely for each model. However, its supported by at least two important and well-written papers that I know of that this parameter can be set automatically:
- https://arxiv.org/abs/1212.0901 "The cutoff threshold for gradient clipping is set based on the average norm of the gradient over one pass on the data"
- https://arxiv.org/abs/1804.09849 "To further stabilize training, we also use adaptive gradient clipping. We discard a training step completely if an anomaly in the gradient norm value is detected, which is usually an indication of an imminent gradient explosion. More specifically, we keep track of a moving average and a moving standard deviation of the log of the gradient norm values, and we abort a step if the norm of the gradient exceeds four standard deviations of the moving average."

In addition, I've found in my own experiments in audio synthesis that setting the parameter automatically is helpful.

## Pitch

Implement a gradient clipping module that stores an internal state used for deciding the threshold for clipping.

**Comment by SsnL** [comment]
I think it is worthwhile to include this in the `contrib` repo.
